src/lib.rs

```diff
@@ -142,21 +142,69 @@ pub struct PrivateKeyRotationCheckResult {
 }
 
 impl PrivateKeyRotationCheckResult {
-    pub fn user_rotation_needed(&self) -> Option<UserId> {
+    pub fn user_rotation_needed(&self) -> Option<&UserId> {
         match &self.rotations_needed {
-            EitherOrBoth::Left(u) | EitherOrBoth::Both(u, _) => Some(u.clone()),
+            EitherOrBoth::Left(u) | EitherOrBoth::Both(u, _) => Some(u),
             _ => None,
         }
     }
 
-    pub fn group_rotation_needed(&self) -> Option<Vec1<GroupId>> {
+    pub fn group_rotation_needed(&self) -> Option<&Vec1<GroupId>> {
         match &self.rotations_needed {
-            EitherOrBoth::Right(groups) | EitherOrBoth::Both(_, groups) => Some(groups.clone()),
+            EitherOrBoth::Right(groups) | EitherOrBoth::Both(_, groups) => Some(groups),
             _ => None,
         }
     }
 
-    // pub fn rotate_all(ironoxide: &IronOxide) -> (UserId, Vec<GroupId>) //TODO consider when group private key rotation is added
+    /// Rotate the private key of the calling user and all groups they are an administrator of where needs_rotation is true.
+    /// Note that this function has the potential to take much longer than other functions, as rotation will be done
+    /// individually on each user/group. If rotation is only needed for a specific group, it is strongly recommended
+    /// to call group_rotate_private_key() instead.
+    pub fn rotate_all(
+        &self,
+        ironoxide: &IronOxide,
+        password: &str,
+    ) -> Result<(
+        Option<UserUpdatePrivateKeyResult>,
+        Option<Vec<GroupUpdatePrivateKeyResult>>,
+    )> {
+        use futures::future::OptionFuture;
+        let user_future = self.user_rotation_needed().and(Some(
+            crate::internal::user_api::user_rotate_private_key(
+                &ironoxide.recrypt,
+                password.try_into()?,
+                ironoxide.device().auth(),
+            ),
+        ));
+        let group_futures = match self.group_rotation_needed() {
```

Comment:
You should be able to map to this instead.

Reply:
I changed this and the user_future both to map to be more readable